src/index/mod.rs

```diff
@@ -0,0 +1,376 @@
+        for (value, keywords) in keywords {
+            for keyword in keywords {
+                modifications
+                    .entry(keyword)
+                    .or_default()
+                    .push((Operation::Addition, value.clone()));
+            }
+        }
+
+        self.findex_graph
+            .insert(self.rng.clone(), &key, modifications, label)
+            .await
+    }
+
+    async fn delete(
+        &mut self,
+        key: &Self::Key,
+        label: &Label,
+        keywords: HashMap<IndexedValue<Keyword, Location>, HashSet<Keyword>>,
+    ) -> Result<HashSet<Keyword>, Self::Error> {
+        let mut seed =
+            <FindexGraph<UserError, EntryTable, ChainTable> as GxEnc<UserError>>::Seed::default();
+        seed.as_mut().copy_from_slice(key.as_bytes());
+        let key = self.findex_graph.derive_keys(&seed);
+
+        let mut modifications = HashMap::<_, Vec<_>>::new();
+        for (value, keywords) in keywords {
+            for keyword in keywords {
+                modifications
+                    .entry(keyword)
+                    .or_default()
+                    .push((Operation::Deletion, value.clone()));
+            }
+        }
+
+        self.findex_graph
+            .insert(self.rng.clone(), &key, modifications, label)
+            .await
+    }
+}
+
+impl<
+    UserError: CallbackErrorTrait,
+    EntryTable: DxEnc<ENTRY_LENGTH, Error = Error<UserError>>
+        + TokenDump<Token = <EntryTable as DxEnc<ENTRY_LENGTH>>::Token, Error = Error<UserError>>,
+    ChainTable: DxEnc<LINK_LENGTH, Error = Error<UserError>>,
+> Findex<UserError, EntryTable, ChainTable>
+{
+    /// Number of items to compact at once.
+    ///
+    /// Given that an entry is EB bytes long and that a link is LB bytes long,
+    /// the memory used by the compact operation is:
+    ///
+    /// N * 32 + BS * EB + f * BS * LB
+    pub const COMPACT_BATCH_SIZE: usize = 1_000_000;
+
+    /// Draw `n` tokens at random among the given `tokens`. The same token may
+    /// be drawn several times, thus the number of tokens returned may be
+    /// lower than `n`.
+    ///
+    /// TODO: update the formula used to select the number of lines to compact.
+    fn select_random_tokens(
+        &self,
+        n: usize,
+        tokens: &[<EntryTable as DxEnc<ENTRY_LENGTH>>::Token],
+    ) -> HashSet<<EntryTable as DxEnc<ENTRY_LENGTH>>::Token> {
+        if tokens.len() <= n {
+            return tokens.iter().cloned().collect();
+        }
+
+        let mut rng = self.rng.lock().expect("could not lock mutex");
+        let mut res = HashSet::with_capacity(n);
+        for _ in 0..n {
+            // In order to draw a random element from the set, draw a random u64 and use it
+            // modulo the length of the set. This is not perfectly uniform but should be
+            // enough in practice.
+            let index = (rng.next_u64() % tokens.len() as u64) as usize;
```

Comment:
iirc, RNG exposes `gen_range(a..b)` method to draw from a uniform distribution.

Reply:
I do not want to depend on the `Rng` trait (which would require compiling the whole `rand` module), but `RngCore` which does not implement `gen_range` (but only requires compiling `rand_core` module).